### PR TITLE
Deck: supporting restrictions on buckets/folders

### PR DESCRIPTION
**Pull request by e-blackwelder**
_Note_: This is a rebase/follow-up to [pr/17650](https://github.com/kubernetes/test-infra/pull/17650), which was originally started by clarketm.

Added _bucket_ and _folder_ validation features to Deck to restrict artifact requests to only specific storage locations. This feature can be enabled by setting `deck.restrict_storage_paths` in the Prow config.

By default, _buckets_ specified in the `plank.decoration_config[*].gcs_configuration.bucket` and 
     the `pr-logs` and `logs` _folders_ are allowed allowed. _Additional buckets_ can be allowed by adding them to 
     `deck.additional_allowed_buckets`. _Additional folders_ can be allowed by adding them to `deck.additional_allowed_folders`.

### CODE REVIEW

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [inline prow/cmd/deck/main.go:953]: I have an idea how to address this, but would rather implement it in a separate patch:
* create an `httpError` type that passes a status code along with the error message (unless we already have that type)
* create a helper method to "getHttpStatusCode" for a given `error` (by default 500, but checking for httpError's status code)
* return an `httpError` in `ValidateStoragePath`
* use the `getHttpStatusCode` method in various places: `handleArtifactView`, etc

Comment (cjwagner) [reply prow/cmd/deck/main.go:953]: That sounds reasonable. Alternatively `renderSpyglass` could just be updated to add an HTTP status code return value.
I think this should be handled in this patch though rather than add new code that we know to be incorrect.

Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: Implemented.

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go:880]: Thanks. I had missed that reference (`ValidateStoragePath` is also called in `job_history.go`). Did a thorough job this time.

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: Done. Skipping the logs for these.

File: prow/cmd/deck/main.go
Comment (alvaroaleman) [inline prow/cmd/deck/main.go:1654]: I do understand the desire to limit the bucket but I am not sure if the added value of verifying the folder warrants the added configuration complexity. If this is really needed (and I can not think of a good reason why, we assume so far that a bucket is a single trust domain) ppl could still configure this via their cloud providers IAM. WDYT?

File: prow/ANNOUNCEMENTS.md
Comment (alvaroaleman) [inline prow/ANNOUNCEMENTS.md:6]: I think it would be good to do this by default

File: prow/cmd/deck/main_test.go
Comment (alvaroaleman) [inline prow/cmd/deck/main_test.go:1059]: Mind testing for the actual error/error message rather than a bool? Tests that use a bool to check for an error tend to rot over time if for some reason a different error is returned

File: prow/ANNOUNCEMENTS.md
Comment (e-blackwelder) [reply prow/ANNOUNCEMENTS.md:6]: Since this is a breaking change, we won't introduce it immediately. I will use a `*bool` for the "turn it on" flag so that we can detect whether it is explicitly set. I will also add warnings to the logs to indicate when the settings need to be updated.

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go:1654]: Per our discussion (with Cole), I will remove folder-level validation, but leave bucket validation. I will instead introduce a file-size validator to fill a similar role.

File: prow/cmd/deck/main_test.go
Comment (e-blackwelder) [reply prow/cmd/deck/main_test.go]: Done.

**Comment by e-blackwelder** [comment]
@alvaroaleman & @cjwagner: There is already a [Deck.Spyglass.SizeLimit](https://github.com/kubernetes/test-infra/blob/e0951ec387d72ee03fd573ca80cbf95c2f74aed7/prow/config/config.go#L616-L621) config field with _some_ validation surrounding it. I'll do a separate PR to address the gaps.

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: FYI: There is already a [Deck.Spyglass.SizeLimit](https://github.com/kubernetes/test-infra/blob/e0951ec387d72ee03fd573ca80cbf95c2f74aed7/prow/config/config.go#L616-L621) config field, with _some_ validation surrounding it. I'll do a separate PR to address artifact size limit gaps.

Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: Done. (Removed the folder-level validation, as requested/discussed.)

File: prow/ANNOUNCEMENTS.md
Comment (e-blackwelder) [reply prow/ANNOUNCEMENTS.md:8]: Done. (Warnings added, "in the future" comments added, etc.)

File: prow/ANNOUNCEMENTS.md
Comment (alvaroaleman) [inline prow/ANNOUNCEMENTS.md:11]: s/allowed allowed/allowed and the folders part isn't accurate anymore

File: prow/cmd/deck/main.go
Comment (alvaroaleman) [inline prow/cmd/deck/main.go:1626]: Why is this exported? It is not used outside of this package or did I miss that?

File: prow/cmd/deck/main_test.go
Comment (alvaroaleman) [inline prow/cmd/deck/main_test.go:1127]: While not technically incorrect its confusing to have part of the testcase in the main test and part in ths sub test, could you move all of it into the func that is called below with `t.Run`?

File: prow/config/config.go
Comment (alvaroaleman) [inline prow/config/config.go:711]: should just be `!RestrictStoragePathsByDefault` but wouldn't it make more sense to warn unconditionally since we want to turn this on by default?

File: prow/ANNOUNCEMENTS.md
Comment (e-blackwelder) [reply prow/ANNOUNCEMENTS.md]: Done. (Thanks for noticing; pretty sure this was an oversight after fixing a merge conflict. I've added a Markdown previewer to my IDE now.)

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: Done. (No, it doesn't seem to be used outside. It has been lowerCamelCased.)

File: prow/cmd/deck/main_test.go
Comment (e-blackwelder) [reply prow/cmd/deck/main_test.go]: Done.

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: I think my intent was to return true *only* when they were restricted _by default_. (I figured if they've already turned off validation when it's off-by-default, they'll leave validation off when it's turned on-by-default.) I added a method comment and made the code more straightforward. Thoughts?

File: prow/cmd/deck/main.go
Comment (alvaroaleman) [inline prow/cmd/deck/main.go:1629]: I don't think this check is correct, for example for a 
* Presubmit, its seven elemtents: https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3840/pull-ci-openshift-installer-master-images/1279069653829685248
* periodic, its four elements: https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3840/pull-ci-openshift-installer-master-images/1279069653829685248
* For a batch, its six elements: https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/batch/pull-ci-integr8ly-integreatly-operator-master-manifest/1279053551028408320

On top of that there are multiple strategies that can be used and that determine the structure on the object storage so IMHO its very dangerous to make any assumptions about that here

File: prow/cmd/deck/main.go
Comment (alvaroaleman) [inline prow/cmd/deck/main.go:1651]: Nit: Why is this named `validate`? Also a nit but I personally find code easier to read without inline funcs because with inline funcs you have to read all the logic rather than being able to say `okay we validate, if there was an error we do X etc`

File: prow/config/config.go
Comment (alvaroaleman) [inline prow/config/config.go:708]: What is the reason of this additional indirection with the constant? Nothing except this func uses the constant or did I miss it? Furthermore I think nothing _should_ use this constant, if anything cares about this setting it should use this func instead.

File: prow/config/config.go
Comment (alvaroaleman) [inline prow/config/config.go:714]: Is there a reason to put this into this func that has exactly one caller rather than doing that check in this exactly one caller? If we want to test the logging, shouldn't we instead just do that rather than testing the func that we think will cause the logging to happen?

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go:708]: I started with just a constant, then refactored in the function later. I think my (vaguely intuitive) "reasons" for leaving both was to make it easier to change-then-delete in the future... `RestrictStoragePathsByDefault` is defined directly beside the `Deck` definition and named extremely similar to the `deck.RestrictStoragePaths` field. Toggling a bool named "ByDefault" would also feel "safer" than re-evaluating the contents of a method like "ShouldValidateSomething". I agree that consumers should be using the function over the constant.

But I'm admittedly _not_ a go-lang expert and I'm guessing/realizing that the additional layers/abstractions aren't idiomatic Go. I'll happily remove them, per your feedback.

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go:1629]: (FYI: This part was originally Travis's change, in case you didn't know/realize.)

I _think_ you're saying that len(parts) should be somewhere between 4 and 7, depending on the type of job, but I'm not sure where you're seeing those numbers (of elements). It appears that the `SplitN(..., 4)` function limits the returned slice to _at most 4_ pieces ([golang docs](https://golang.org/pkg/strings/#SplitN)), meaning the `<sub-folder>` piece is just checked for existence.

That being said, I've personally had a bit of a problem with the pseudo-URL format of the `path` parameter -- I would have expected a "://" after the "<key-type>" (as opposed to a single "/"). I had previously noticed at least one caller (possibly the sole caller) is string-formatting the path from various pieces. I think I'd rather just pass and validate in the path as a struct, which would mean more code, but I think we'd both be satisfied.

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go:714]: As far as "reasons"... just my personal experience/style, I'd say. My preference is usually something resembling "single level of abstraction per function". Because (for example), it's easier to write confidence-building unit-tests for tiny methods like this vs "did Xxxlogging framework get called with the right varargs". But (as I mention above), I'll defer to your Go-lang sensibilities. :)

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go:1651]: Sorry, refactoring noise. I'll take another pass. (Not quite satisfied with where I left this, either.)

File: prow/cmd/deck/main.go
Comment (alvaroaleman) [inline prow/cmd/deck/main.go:1639]: There is another source of valid buckets we have to consider: They can be set on jobs. So we need to walk all jobs, check if they have `.DecorationConfig != nil && .DecorationConfig.GCSConfiguration != nil && .DecorationConfig.GCSConfiguration.BucketName != "`.

Maybe it would be better to construct that set in `Load` and store it somewhere on the config?

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: I looked into the [Path Strategies](https://github.com/kubernetes/test-infra/blob/088d21b2a38643ff55e59069228f449dacd4522e/prow/apis/prowjobs/v1/types.go#L529-L531) and they do have a few different ways of formatting paths. Since I'm only concerned with bucket names now, they can all be validated in the same way -- which is just "is there a folder".

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: Removed the extra `RestrictStoragePathsByDefault` const.

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: Removed extraneous function `ShouldWarnAboutInvalidStoragePaths`.

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: Done.

_Note_: This can't handle buckets defined on jobs via InRepoConfig. For now (until ~Jan 2021), I'm skipping validation on InRepoConfig-enabled Prows and providing a detailed warning on how to prepare.

**Comment by e-blackwelder** [comment]
> So we do not do size validation for now?

See https://github.com/kubernetes/test-infra/pull/19250

File: prow/cmd/deck/main_test.go
Comment (cjwagner) [inline prow/cmd/deck/main_test.go:1171]: nit: A slice of testcase structs `{err error, expectedCode int}` and a loop would be a bit more idiomatic and less repetitive, but it doesn't really matter here.

File: prow/config/config.go
Comment (cjwagner) [inline prow/config/config.go:744]: "the built-in and default storage locations" is not clear to me. Please elaborate here or link/reference documentations that does.

File: prow/config/config.go
Comment (cjwagner) [inline prow/config/config.go:754]: nit: You don't need to use `strings.Join` here. You can
1. use ``` ` ``` for multiline raw strings
1. concatenate lines with `+` (the compiler should be smart enough to do this for literals at compile time rather than runtime)
1. use a long single line literal since this is for an error message.

I think those options are more efficient and more readable.

File: prow/config/config.go
Comment (cjwagner) [inline prow/config/config.go:769]: https://godoc.org/github.com/kubernetes/apimachinery/pkg/util/sets#String.List

File: prow/config/config.go
Comment (petr-muller) [inline prow/config/config.go:762]: nit: move this immediately before the usage, after the `ShouldValiate...` check which  does not use this.

File: prow/logrusutil/logrusutil.go
Comment (e-blackwelder) [reply prow/logrusutil/logrusutil.go]: Done.

File: prow/plugins/config.go
Comment (e-blackwelder) [reply prow/plugins/config.go]: Done (you troll :P).

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: Done.

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: Done.

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: Done.

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: Done.

**Review by e-blackwelder — COMMENTED** [review]
/hold

Would like to wait for Alvaro to review. (And still have a TODO from Cole.)

File: prow/config/config.go
Comment (e-blackwelder) [reply prow/config/config.go]: Done.

File: prow/cmd/deck/main.go
Comment (e-blackwelder) [reply prow/cmd/deck/main.go]: Done. Moved validation into `newBlobStorageBucket` (job_history.go) and `StorageArtifactFetcher.newStorageJobSource` (with minor refactoring to make it work).

Balancing URL-formatting (gs/bucket/folder vs gs://bucket/folder) and bucket-vs-full-path was tricky, but method names, param names, and comments should fully reflect expectations for each.

**Comment by e-blackwelder** [comment]
> Would like to wait for Alvaro to review. (And still have a TODO from Cole.)

Finished final TODO from Cole's review.

File: prow/spyglass/spyglass.go
Comment (cjwagner) [inline prow/spyglass/spyglass.go:88]: We should pass a dynamic config getter here rather than a method of the static config struct. As is, if a new job is added with a new bucket, we'd have to restart Deck in order for it to respect the newly valid bucket.

Comment (e-blackwelder) [reply prow/spyglass/spyglass.go]: Done.

**Review by cjwagner — APPROVED** [review]

**Pull request merged**
